### DISCUSSION TITLE
Return value of Spatie\Once\Backtrace::getObject() must be an instance of Spatie\Once\mixed, instance of Laravel\Nova\Http\Requests\NovaRequest returned

### DISCUSSION THREAD

**Discussion by andresilvagomez** [laravel/nova-issues#3187] [Q&A] [answered]
- Laravel Version: 8.22.1
- Nova Version: 3.19.1
- PHP Version: 7.4
- Database Driver & Version: Postgres
- Operating System and Version: #.#
- Browser type and version: #.#
- Reproduction Repository: https://github.com/###/###

### Description:

Return value of Spatie\Once\Backtrace::getObject() must be an instance of Spatie\Once\mixed, instance of Laravel\Nova\Http\Requests\NovaRequest returned

**Comment by FaridAghili** [comment, marked as answer]
This happened to me when I was developing on my local machine with PHP 8 installed, and deployed the project to shared hosting which had PHP 7.4 installed.
Notice that, I copied whole Laravel project, including `vendor` folder to remote host and didn't run `composer install` on the remote host.

In fact, on my local machine `spatie/once` v3 got installed by composer (which requires PHP 8), and PHP 8 wasn't available on my hosting.

If you're doing the same, I suggest to install PHP 7.* on your local machine, run `composer update`, then deploy to server.

**Comment by crynobone** [comment]
Or follow this step https://twitter.com/themsaid/status/1341715935741284353?s=20

**Comment by andresilvagomez** [comment]
@FaridAghili thanks a lot

**Comment by seobrain** [comment]
I can confirm that this issue appears on php 7.4, but not on php 8.0.

I have latest versions of Laravel and Nova.

In order to reproduce just install empty Laravel project from scratch and add Nova via Composer.

Hope it helps.

**Reply by henryavila** [reply]
> I can confirm that this issue appears on php 7.4, but not on php 8.0.
> 
> I have latest versions of Laravel and Nova.
> 
> In order to reproduce just install empty Laravel project from scratch and add Nova via Composer.
> 
> Hope it helps.

This error happen when you try to use spatie/once v3 on a php 7.4 environment. (It happened with me too)

As pointed in v3 release: https://github.com/spatie/once/releases/tag/3.0.0
The v2 drop support for php 7.

So the options are:
- Upgrade to php 8 OR
- Use spatie/once v2.x

Look at your composer.json if you dont manually add `spatie/once ^3.x`.
Laravel nova already has the dependency as `"spatie/once": "^1.1|^2.0|^3.0",` , so you don't need to add it to your composer file.

If in your local environment you use php 8 the composer will install spatie/once v3, but if deploy to a server with php 7 it will lead to a error. In this case, just add to you `composer.json`: `"spatie/once": "^2.0"`

**Comment by crynobone** [comment]
@seobrain unable to reproduce.

https://user-images.githubusercontent.com/172966/104862839-2a3fd380-596f-11eb-8b66-85c514ba31d3.mp4

```bash
function replicate-nova-issue() {
    valet use php@7.4
    cd ~/Projects/laravel/nova/issues
    laravel new issue-3170
    cd issue-3170
    composer config repositories.local '{"type": "composer", "url": "https://nova.laravel.com"}' --file composer.json
    composer require "laravel/nova:*"
    composer show | grep "spatie/once"
}
```
